driver/csiplugin/controllerserver.go

```diff
@@ -1818,3 +1827,108 @@ func (cs *ScaleControllerServer) ControllerExpandVolume(ctx context.Context, req
 func (cs *ScaleControllerServer) ControllerGetVolume(ctx context.Context, req *csi.ControllerGetVolumeRequest) (*csi.ControllerGetVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "")
 }
+
+// getRemoteClusterID returns the cluster ID for the passed cluster name.
+func (cs *ScaleControllerServer) getRemoteClusterID(clusterName string) (string, error) {
+
+	glog.V(5).Infof("fetching cluster details from cache map for cluster %s", clusterName)
+	clusterDetails, found := cs.Driver.clusterMap.Load(ClusterName{clusterName})
+	if found {
+		glog.V(5).Infof("cluster details found from cache map for cluster %s", clusterName)
+		updateTime := clusterDetails.(ClusterDetails).lastupdated
+		expiryDuration := clusterDetails.(ClusterDetails).expiryDuration
+		glog.V(5).Infof("checking if cluster details found from cache map for cluster %s has expired.", clusterName)
+		if time.Since(updateTime).Hours() < float64(expiryDuration) { // cluster details are not expired
+			glog.V(5).Infof("cluster details found from cache map for cluster %s are valid.", clusterName)
+			return clusterDetails.(ClusterDetails).id, nil
+		} else { // cluster details are expired
+			glog.V(5).Infof("cluster details found from cache map for cluster %s are expired.", clusterName)
+			cID := clusterDetails.(ClusterDetails).id
+			conn, err := cs.getConnFromClusterID(cID)
+			if err != nil {
+				return "", err
+			}
+			clusterSummary, err := conn.GetClusterSummary()
+			if err != nil {
+				return "", err
+			}
+			cName := clusterSummary.ClusterName
+			if cName == clusterName {
+				glog.V(5).Infof("updating cluster details in cache map for cluster %s.", clusterName)
+				cs.Driver.clusterMap.Store(ClusterName{cName}, ClusterDetails{cID, cName, time.Now(), 24})
+				cs.Driver.clusterMap.Store(ClusterID{cID}, ClusterDetails{cID, cName, time.Now(), 24})
+				return cID, nil
+			} else {
+				found = false
+			}
+		}
+	}
+
+	if !found {
+		glog.V(5).Infof("cluster details are either expired or not found in cache map for cluster %s. Updating the cache map.", clusterName)
+		scaleconfig := settings.LoadScaleConfigSettings()
+
+		for i := range scaleconfig.Clusters {
+
+			var cID string
+			var cName string
+
+			glog.V(5).Infof("creating new connector for the cluster %s", scaleconfig.Clusters[i].ID)
+			clusterConnector, err := connectors.NewSpectrumRestV2(scaleconfig.Clusters[i])
+			if err != nil {
+				glog.V(5).Infof("unable to create new connector for the cluster %s", scaleconfig.Clusters[i].ID)
+				continue
+			}
+
+			cID = scaleconfig.Clusters[i].ID
+			glog.V(5).Infof("fetching cluster details from cache map for cluster %s", scaleconfig.Clusters[i].ID)
+			clusterDetails, found := cs.Driver.clusterMap.Load(ClusterID{cID})
+			if found {
+				glog.V(5).Infof("checking if cluster details found from cache map for cluster %s has expired.", scaleconfig.Clusters[i].ID)
```

Comment:
The lines of code starting here for "if found" part look similar/same as code above starting at line 1837. Can we have a function for these common lines of code ? Future maintenance of code becomes easy for any programmer.

Reply:
created function `checkExpiry()` and `updateClusterMap()` to handle this. please review.

Reply:
created function `checkExpiry()` and `updateClusterMap()` to handle this. please review.